agent/testflinger_agent/tarfile_patch.py

```diff
@@ -0,0 +1,260 @@
+try:
+    # attempt importing a filter, to check if the feature is present
+    from tarfile import data_filter
+except ImportError:
```

Comment:
This "file-level" try..except makes everything indented. It's not like we're running out of storage space for 0x20 bytes, but maybe this try..except should be done in `agent.py`, like this:

```python3
try:
    from tarfile import data_filter
    import tarfile
    del data_filter # optional, if you want to clean up after the try-import
except ImportError:
    from . import tarfile_patch as tarfile
```

Then this file (`tarfile_patch.py`) can just contain the implementation, and the "policy" whether to use the Python-supplied `tarfile` module or `tarfile_patch` is up to the user.

Alternatively:

```python3
import tarfile

if not hasattr(tarfile, 'data_filter'):
    # Python < 3.8.17 compatibility
    import tarfile_patch
    tarfile.open = tarfile_patch.open
```

Or even:

```python3
import sys

if sys.version_info < (3, 12):
    import .tarfile_patch as tarfile
else:
    import tarfile
```

(the versions and names might be slightly different, haven't checked)

Reply:
I moved the conditional import logic into the agent, you are absolutely right that it belongs there.